File: website/serve.pl
#!/usr/bin/perl
use strict;
use warnings;
use IO::Socket::INET;
use File::Basename;
use POSIX qw(strftime);

my $port = 3456;
my $root = '/Users/connorblades32/Documents/Claude/Worksop Workspace';

my %mime = (
    html => 'text/html',
    css  => 'text/css',
    js   => 'application/javascript',
    png  => 'image/png',
    jpg  => 'image/jpeg',
    jpeg => 'image/jpeg',
    gif  => 'image/gif',
    svg  => 'image/svg+xml',
    ico  => 'image/x-icon',
    woff2=> 'font/woff2',
    woff => 'font/woff',
);

my $server = IO::Socket::INET->new(
    LocalAddr => '127.0.0.1',
    LocalPort => $port,
    Proto     => 'tcp',
    Listen    => 10,
    ReuseAddr => 1,
) or die "Cannot start server: $!";

print "Server running on http://127.0.0.1:$port\n";

while (my $client = $server->accept()) {
    my $request = '';
    while (my $line = <$client>) {
        $request .= $line;
        last if $line eq "\r\n";
    }

    my ($method, $path) = $request =~ /^(\w+)\s+(\S+)/;
    $path = '/index.html' if !$path || $path eq '/';
    $path =~ s/\?.*//;
    $path =~ s/\.\.//g;

    my $file = $root . $path;
    my ($ext) = $file =~ /\.(\w+)$/;
    $ext = lc($ext // 'html');
    my $type = $mime{$ext} // 'application/octet-stream';

    if (-f $file) {
        open(my $fh, '<:raw', $file) or do {
            print $client "HTTP/1.1 403 Forbidden\r\nContent-Length: 9\r\n\r\nForbidden";
            close $client;
            next;
        };
        my $content = do { local $/; <$fh> };
        close $fh;
        my $len = length($content);
        print $client "HTTP/1.1 200 OK\r\nContent-Type: $type\r\nContent-Length: $len\r\nConnection: close\r\n\r\n$content";
    } else {
        print $client "HTTP/1.1 404 Not Found\r\nContent-Length: 13\r\n\r\n404 Not Found";
    }
    close $client;
}

File: website/serve.js
const http = require('http');
const fs   = require('fs');
const path = require('path');

const PORT = process.env.PORT || 8765;
const ROOT = __dirname;

const MIME = {
  '.html': 'text/html', '.css': 'text/css', '.js': 'application/javascript',
  '.mp4': 'video/mp4', '.png': 'image/png', '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg', '.svg': 'image/svg+xml', '.ico': 'image/x-icon',
  '.webp': 'image/webp', '.woff2': 'font/woff2', '.woff': 'font/woff',
};

http.createServer((req, res) => {
  let urlPath = req.url.split('?')[0];
  if (urlPath === '/') urlPath = '/index.html';
  else if (!path.extname(urlPath)) urlPath = urlPath.replace(/\/$/, '') + '/index.html';

  const filePath = path.join(ROOT, urlPath);
  const ext = path.extname(filePath);

  fs.stat(filePath, (err, stat) => {
    if (err) { res.writeHead(404); res.end('Not found'); return; }

    const mime = MIME[ext] || 'application/octet-stream';
    const rangeHeader = req.headers.range;

    /* ── Range requests (required for video seeking) ── */
    if (rangeHeader && ext === '.mp4') {
      const fileSize = stat.size;
      const [startStr, endStr] = rangeHeader.replace(/bytes=/, '').split('-');
      const start = parseInt(startStr, 10);
      const end   = endStr ? parseInt(endStr, 10) : fileSize - 1;
      const chunk = end - start + 1;

      res.writeHead(206, {
        'Content-Range':  `bytes ${start}-${end}/${fileSize}`,
        'Accept-Ranges':  'bytes',
        'Content-Length': chunk,
        'Content-Type':   mime,
      });
      fs.createReadStream(filePath, { start, end }).pipe(res);
      return;
    }

    /* ── Normal full-file response ── */
    res.writeHead(200, {
      'Content-Type':   mime,
      'Content-Length': stat.size,
      'Accept-Ranges':  'bytes',
    });
    fs.createReadStream(filePath).pipe(res);
  });

}).listen(PORT, () => console.log(`Server running on http://localhost:${PORT}`));
